### PR TITLE
Introduce expandStringAsync and expandStringSync in url-replacement

### PR DESCRIPTION
**Pull request by mkhatib**
Follow up to #6257

Related to: #6130 and #5654

### CODE REVIEW

File: build-system/tasks/presubmit-checks.js [hunk 1]
```diff
@@ -390,6 +390,19 @@ var forbiddenTerms = {
       'src/service/url-replacements-impl.js',
     ]
   },
+  'expandStringSync\\W': {
```
Comment (jridgewell) [inline build-system/tasks/presubmit-checks.js]: These should be in `forbiddenTermsSrcInclusive`. Also: `\\.expandStringSync\\(`

File: src/service/url-replacements-impl.js [hunk 2]
```diff
@@ -532,21 +534,85 @@ export class UrlReplacements {
    * @return {string}
    */
   expandSync(url, opt_bindings, opt_collectVars, opt_whiteList) {
-    return /** @type {string} */(
-        this.expand_(url, opt_bindings, opt_collectVars, /* opt_sync */ true,
-            opt_whiteList));
+    return /** @type {string} */ (this.expandUrlSync(
```
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: Useless typecast?

File: src/service/url-replacements-impl.js [hunk 3]
```diff
@@ -532,21 +534,85 @@ export class UrlReplacements {
...
    * or override existing ones.
+   *
+   * TODO(mkhatib, #6322): Deprecate and please use expandUrlAsync or expandStringAsync.
    * @param {string} url
    * @param {!Object<string, *>=} opt_bindings
    * @return {!Promise<string>}
    */
   expandAsync(url, opt_bindings) {
-    return /** @type {!Promise<string>} */(this.expand_(url, opt_bindings));
+    return /** @type {!Promise<string>} */ (
```
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: Useless typecast?

File: src/service/url-replacements-impl.js [hunk 4]
```diff
@@ -532,21 +534,85 @@ export class UrlReplacements {
...
+   * variables or override existing ones.  Any async bindings are ignored.
+   * @param {string} url
+   * @param {!Object<string, (ResolverReturnDef|!SyncResolverDef)>=} opt_bindings
+   * @param {!Object<string, ResolverReturnDef>=} opt_collectVars
+   * @param {!Object<string, boolean>=} opt_whiteList Optional white list of names
+   *     that can be substituted.
+   * @return {string}
+   */
+  expandUrlSync(url, opt_bindings, opt_collectVars, opt_whiteList) {
+    return /** @type {string} */ (
```
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: Useless typecast?

File: build-system/tasks/presubmit-checks.js (hunk 1 above)
Comment (mkhatib) [inline build-system/tasks/presubmit-checks.js]: Will do, what is the difference between the `forbiddenTermsSrcInclusive` and `forbiddenTerms`?

Comment (mkhatib) [inline build-system/tasks/presubmit-checks.js]: Done

File: src/service/url-replacements-impl.js (hunk 2 above)
Comment (mkhatib) [inline src/service/url-replacements-impl.js]: Done

File: src/service/url-replacements-impl.js (hunk 3 above)
Comment (mkhatib) [inline src/service/url-replacements-impl.js]: Done

File: src/service/url-replacements-impl.js (hunk 4 above)
Comment (mkhatib) [inline src/service/url-replacements-impl.js]: Done

File: build-system/tasks/presubmit-checks.js (hunk 1 above)
Comment (jridgewell) [inline build-system/tasks/presubmit-checks.js]: Test code is allowed to use it in `forbiddenTermsSrcInclusive`.

File: src/service/url-replacements-impl.js [hunk 5]
```diff
@@ -532,21 +534,83 @@ export class UrlReplacements {
...
+   * variables or override existing ones.  Any async bindings are ignored.
+   * @param {string} url
+   * @param {!Object<string, (ResolverReturnDef|!SyncResolverDef)>=} opt_bindings
+   * @param {!Object<string, ResolverReturnDef>=} opt_collectVars
+   * @param {!Object<string, boolean>=} opt_whiteList Optional white list of names
+   *     that can be substituted.
+   * @return {string}
+   */
+  expandUrlSync(url, opt_bindings, opt_collectVars, opt_whiteList) {
+    return this.ensureProtocolMatches_(url, /** @type {string} */ (this.expand_(
```
Comment (dvoytenko) [inline src/service/url-replacements-impl.js]: Do we need to do the same for the whole origin?

**Comment by dvoytenko** [comment]
LGTM with one question

File: src/service/url-replacements-impl.js (hunk 5 above)
Comment (mkhatib) [inline src/service/url-replacements-impl.js]: Not sure what do you mean, /cc @jridgewell who worked on this.

Comment (mkhatib) [inline src/service/url-replacements-impl.js]: Oh I see what you mean, basically ensure that the origin hasn't changed. @jridgewell what was the reasoning behind protecting protocol changes and does it apply to origin changes?

Comment (mkhatib) [inline src/service/url-replacements-impl.js]: Also, if you guys don't mind I'll go ahead and merge this and I opened #6338 for continuing this discussion.

**Pull request merged**